src/main/java/org/opensearch/flowframework/rest/AbstractSearchWorkflowAction.java

```diff
@@ -0,0 +1,115 @@
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.flowframework.common.FlowFrameworkFeatureEnabledSetting;
+import org.opensearch.flowframework.exception.FlowFrameworkException;
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.BytesRestResponse;
+import org.opensearch.rest.RestChannel;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.rest.RestResponse;
+import org.opensearch.rest.action.RestResponseListener;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opensearch.core.xcontent.ToXContent.EMPTY_PARAMS;
+import static org.opensearch.flowframework.common.FlowFrameworkFeatureEnabledSetting.FLOW_FRAMEWORK_ENABLED;
+import static org.opensearch.flowframework.util.RestHandlerUtils.getSourceContext;
+
+/**
+ * Abstract class to handle search request.
+ */
+public abstract class AbstractSearchWorkflowAction<T extends ToXContentObject> extends BaseRestHandler {
+
+    protected final List<String> urlPaths;
+    protected final String index;
+    protected final Class<T> clazz;
+    protected final ActionType<SearchResponse> actionType;
+    protected final FlowFrameworkFeatureEnabledSetting flowFrameworkFeatureEnabledSetting;
+
+    /**
+     * Instantiates a new AbstractSearchWorkflowAction
+     * @param urlPaths urlPaths to create routes
+     * @param index index the search should be done on
+     * @param clazz model class
+     * @param actionType from which action abstract class is called
+     * @param flowFrameworkFeatureEnabledSetting Whether this API is enabled
+     */
+    public AbstractSearchWorkflowAction(
+        List<String> urlPaths,
+        String index,
+        Class<T> clazz,
+        ActionType<SearchResponse> actionType,
+        FlowFrameworkFeatureEnabledSetting flowFrameworkFeatureEnabledSetting
+    ) {
+        this.urlPaths = urlPaths;
+        this.index = index;
+        this.clazz = clazz;
+        this.actionType = actionType;
+        this.flowFrameworkFeatureEnabledSetting = flowFrameworkFeatureEnabledSetting;
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        if (!flowFrameworkFeatureEnabledSetting.isFlowFrameworkEnabled()) {
+            FlowFrameworkException ffe = new FlowFrameworkException(
+                "This API is disabled. To enable it, update the setting [" + FLOW_FRAMEWORK_ENABLED.getKey() + "] to true.",
+                RestStatus.FORBIDDEN
+            );
+            return channel -> channel.sendResponse(
+                new BytesRestResponse(ffe.getRestStatus(), ffe.toXContent(channel.newErrorBuilder(), ToXContent.EMPTY_PARAMS))
```

Comment:
just wondering what is the toXContent adding here, I just see: `return builder.startObject().field("error", this.getMessage()).endObject();` does it just add error before message?

Reply:
Yes, it just converts the exception to json xcontent so that we can write it back to the rest channel

Reply:
Discussed more offline with Josh and its the current best way we have for converting the exception to JSON when we want to take control of what the status will be. Can't think of a better way though, it just seemed like adding more work to call on ToXContent each time since we didn't have that in AD. Understand why its done now a little better